### ISSUE TITLE
AWS ECS agent does not start in EC2 instance

### ISSUE THREAD

**Issue by tscodeler** [aws/amazon-ecs-agent#4130]
### Summary
AWS ECS agent does not start in EC2 instance

### Description
It looks like there might be an issue with the ECS agent on my ECS cluster. For the past two weeks, my ECS cluster with EC2 instances managed by auto scaling (launch templates) and capacity provider has been working fine. However, new instances are not being connected to the ECS cluster because the agent is not starting anymore.

Even when I try to start the ECS agent manually on the instance, it hangs.

The Docker service is running properly, and the proper ECS role is attached to the instance. There are no logs for the agent on the instance.

The AMI I'm using is "amzn2-ami-ecs-hvm-2.0.20240319-x86_64-ebs" with the ID "ami-06ebbcdf40f9949e7." Already tried some new AMI versions but facing the same issue.

Here's the ECS service status on a freshly launched EC2 instance:

```
ecs.service - ECS Agent
   Loaded: loaded (/usr/lib/systemd/system/ecs.service; enabled; vendor preset: disabled)
   Active: inactive (dead)
```

### Expected Behavior
ECS agent service starting automatically.

### Environment Details
* docker info:
```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc., v0.0.0+unknown)

Server:
 Containers: 2
  Running: 2
  Paused: 0
  Stopped: 0
 Images: 5
 Server Version: 20.10.25
 Storage Driver: overlay2
  Backing Filesystem: xfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 64b8a811b07ba6288238eefc14d898ee0b5b99ba
 runc version: 4bccb38cc9cf198d52bebf2b3a90cd14e7af8c06
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 4.14.336-257.566.amzn2.x86_64
 Operating System: Amazon Linux 2
 OSType: linux
 Architecture: x86_64
 CPUs: 2
 Total Memory: 14.91GiB
 Name: ip-10-0-4-66.us-west-2.compute.internal
 ID: Q2XT:HGAQ:XXEK:T7OZ:7C2Y:WJYW:44JQ:OWYE:VUG5:FSOB:QBAV:MPPK
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

* curl http://localhost:51678/v1/metadata
```
"Version":"Amazon ECS Agent - v1.82.1
```

### Supporting Log Snippets
* journalctl -xeu ecs.service
```
---no entries---
```

**Comment by hozkaya2000** [comment]
Hello Thiago, 

Has anything changed since you posted this last week? Unfortunately, we can't diagnose the issue without any logs. For further investigation, we suggest using [amazon-ecs-logs-collector](https://github.com/aws/amazon-ecs-logs-collector) and providing us with the logs. For a temporary mitigation, we suggest you try a few older AMI versions (since the new ones do not work based on your previous attempts)

**Comment by thiagoscodelerae** [comment]
@hozkaya2000  thank you for your answer. I'll take a look at `amazon-ecs-logs-collector` . Regardless of the AMI version, the agent isn't working now, even though it was fine earlier. I have tried using an AMI that worked previously, but it is still not working.

**Comment by cagejsn-importal** [comment]
hi @thiagoscodelerae , i am having the exact same issue I think, were you able to find a solution?

**Comment by thiagoscodelerae** [comment]
@cagejsn-importal Hi, you can find my workaround for that (instance reboot) and another person's workaround here: https://stackoverflow.com/questions/78244716/aws-ecs-agent-does-not-start-in-ec2-instance

**Comment by ofiry-tm** [comment]
I have the same issue...

**Comment by chrisella** [comment]
Thanks @garysferrao , I've just followed those simple steps and immediately the instance appears in my ECS cluster availability.
I did that all manually connected via SSM so now need to see if I can automate this so future instances work because otherwise this isn't all that handy for our auto scaling group setup.

**Comment by thiagoscodelerae** [comment]
@chrisella maybe you could automate this by adding the cloud-final.service restart command as part of the user data and using a launch template for your instances.

**Comment by sharuzzaman** [comment]
I'm facing this issue today and also seeing cloud-final.service is stuck.

ECS agent successfully running after restarting cloud-final.service

AMI ID: ami-0db7668cca2268dd3
AMI name: al2023-ami-ecs-hvm-2023.0.20251002-kernel-6.1-x86_64
Region: ap-southeast-5

**Comment by sparrc** [comment]
Yes, AL2 is being deprecated on June 30, 2026. This issue is getting a bit hard to track what the real issue is, since we have reports from both AL2 and AL2023.

I'm going to close this issue, please feel free to open a new issue if you have a specific problem with agent starting on an instance. As @prateekchaudhry noted, restarting ecs agent in EC2 userdata will not work.

If you face this issue we would need as many system/agent logs as possible to debug, so please consider running the ecs logs collector: https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-logs-collector.html

**Comment by garysferrao** [comment]
@sparrc 

> As prateekchaudhry noted, restarting ecs agent in EC2 userdata will not work.

sorry, i'm not sure you understood my comment. i meant that it was the same problem even before restarting the ecs agent, because i wouldn't arbitrarily do it if it wasn't a problem in the first place.
and again, another person commented the same even in the AL2023 AMI.
do i have to reproduce it again and share the logs?
